src/perf_counters.cc

```diff
@@ -32,17 +32,22 @@ const bool PerfCounters::kSupported = true;
 
 bool PerfCounters::Initialize() { return pfm_initialize() == PFM_SUCCESS; }
 
-PerfCounters PerfCounters::Create(
+PerfCounters* PerfCounters::Create(
     const std::vector<std::string>& counter_names) {
+  static std::unique_ptr<PerfCounters> perf_counters_ptr;
```

Comment:
How about this instead, so we also exit cleanly (i.e. cleaning up file descriptors):
- don't have the statics here (lines 37-38)
- add a `int ref_count_ =1` field to `PerfCounters`, and a `void refcount() { ++ref_count; }`
- `PerfCountersMeasurement`:
   - has a static `std::unique_ptr<PerfCounters> counters_` field.
   - its ctor takes a `const std::vector<std::string>& counter_names`. It checks if `counters_` is nullptr, if so, creates it with the given `counter_names` and returns its contained ptr. If it's not nullptr, then sanity check the `counter_names` match what we have in `counters_`, and `counters_->refcount(); return counters_.get()`.
   -  the dtor of `PerfCounters` decrements `ref_count_`, and when that hits 0, closes the file descriptors as it does today

Add a comment on `PerfCountersMeasurement` that the ctor is not threadsafe (which should be fine, the scenario is that these are create upfront, at set up); and that the string vector must match what's cached (which is both fine and unavoidable, the whole point is that we can't have too many simultaneously `perf_event_open`-ed counters.)

Reply:
OK, but this still relies on process shutdown to clean up everything, which is not quite best-practice.

I fixed my earlier comment above, it wasn't quite right; also, if thread safety is a concern, it can easily be mitigated there by having a mutex for operations on the shared `counters` and the `ref_count_` field.

Reply:
Implemented this in be1b7.
Mutex will be added soon (some bugs appeared).